### PR TITLE
RavenDB-14554 - Release large contexts and when old ones when we are in low memory state

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-14554

### CODE REVIEW

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs [hunk 1]
```diff
@@ -32,6 +32,12 @@ public MemoryConfiguration()
         [ConfigurationEntry("Memory.LowMemoryCommitLimitInMb", ConfigurationEntryScope.ServerWideOnly)]
         public Size LowMemoryCommitLimitInMb { get; set; }
 
+        [Description("The maximum size of context to keep in the pool after dispose")]
+        [DefaultValue(32)]
+        [SizeUnit(SizeUnit.Megabytes)]
+        [ConfigurationEntry("Memory.MaxContextSizeToKeep", ConfigurationEntryScope.ServerWideOnly)]
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/MemoryConfiguration.cs]: Should be `Memory.MaxContextSizeToKeepInMb`

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 2]
```diff
@@ -417,7 +435,10 @@ public virtual void Dispose()
         public void LowMemory()
         {
             if (LowMemoryFlag.Raise())
+            {
+                Generation++;
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: Safer maybe to do Interlocked.Increment?

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 3]
```diff
@@ -156,12 +158,14 @@ private void DisposeOfContexts()
             }
         }
 
-        protected JsonContextPoolBase()
+        protected JsonContextPoolBase(Size? maxContextSizeToKeep)
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: create two separate ctors to not break binary compatibility

File: src/Sparrow/Json/JsonContextPool.cs [hunk 4]
```diff
@@ -2,6 +2,10 @@
 {
     public class JsonContextPool : JsonContextPoolBase<JsonOperationContext>
     {
+        public JsonContextPool(Size? maxContextSizeToKeep = null) : base(maxContextSizeToKeep)
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonContextPool.cs]: create two separate ctors to not break binary compatibility

**Comment by ppekrol** [comment]
I've rebased your PR.

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Config/Categories/MemoryConfiguration.cs]: done

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: done

File: src/Sparrow/Json/JsonContextPool.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Sparrow/Json/JsonContextPool.cs]: done

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 2 above)
Comment (ayende) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: Yes, that would be good to do

**Review by ayende — APPROVED** [review]

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: done

**Pull request merged**